src/Functions/FunctionsPunycode.cpp

```diff
@@ -0,0 +1,329 @@
+
+                    std::copy_n(acePrefix.data(), acePrefix.size(), encoded_string + encoded_length);
+                    encoded_length += acePrefix.size();
+                    std::copy_n(buf, length, encoded_string + encoded_length);
+                    encoded_length += length;
+                }
+                else
+                {
+                    encoded_length += appendUTF32ToString(input.data(), prev, idx, encoded_string + encoded_length);
+                }
+
+                encoded_string[encoded_length++] = '.';
+                // To skip '.' with prefix increment '++idx'
+                prev = ++idx;
+            }
+        }
+
+        if (!isASCIIStrUTF32(input, prev, input.size()))
+        {
+            const auto result_code = punycode_encode(input.length() - prev, input.data() + prev,
+                                                     nullptr, &(length = labelMaxLength), buf);
+            if (punycode_success != result_code)
+                return 0;
+
+            std::copy_n(acePrefix.data(), acePrefix.size(), encoded_string + encoded_length);
+            encoded_length += acePrefix.size();
+            std::copy_n(buf, length, encoded_string + encoded_length);
+            encoded_length += length;
+        }
+        else
+        {
+            encoded_length += appendUTF32ToString(input.data(), prev, input.size(), encoded_string + encoded_length);
+        }
+        return encoded_length;
+    }
+
+    /** If the string contains ' :// ' then this string contains the full DNS name
+     *  if the string do not contains ' :// ' then we consider the entire input
+     *  as data to be encoded regardless of the content of characters in it ' / '
+     */
+    constexpr std::pair<size_t, size_t> getFqdnBounds(std::string_view input) noexcept{
+        std::string::size_type start = input.find("://");
+        if (std::string::npos == start)
+            return {0, input.size() };
+        std::string::size_type last = input.find('/', start + 3);
+        return {start + 3, std::string::npos == last ? input.size() : last};
+    };
+
+    size_t punycodeEncode(const std::string_view input,
+                          UInt8 * encoded_string)
+    {
+        const auto [start, last] = getFqdnBounds(input);
+        std::copy_n(input.data(), start, encoded_string);
+
+        auto s32 = converter.from_bytes((input.data() + start), (input.data() + last));
+        std::for_each(s32.begin(), s32.end(), [](auto & ch) { ch = Poco::Unicode::toLower(ch);});
+        size_t decoded_length = punycodeEncodeInternal(s32, encoded_string + start);
+        if (0 == decoded_length)
+            throw std::runtime_error("Failed to encode");
+        decoded_length += start;
+        if (input.size() > last)
+            encoded_string[decoded_length++] = '/';
+        return decoded_length;
+    }
+
+    size_t punycodeDecode(const std::string_view input,
+                          UInt8 * decoded_string)
```

Comment:
It's better and more safe to pass `ColumnString::Chars & res_data` to this function instead of `UInt8 * decoded_string` because this way you can just push back new characters to `res_data` without thinking about how much memory you need to reserve before.